src/components/Tooltip/TooltipPosition.js

```diff
@@ -0,0 +1,164 @@
+    }
+
+    const arrowPosition = {
+      top: 0,
+      left: 0,
+      placement: placement
+    }
+
+    if (['top','bottom'].includes(placement)) {
+      // Center it
+      const targetCenterX  = targetRect.width / 2
+      overlayRect.left = targetRect.left + targetCenterX - (overlayRect.width / 2)
+      const overlayDistanceFromRightEdge = overlayRect.left + overlayRect.width
+
+      if (placement === 'top') {
+        overlayRect.top = targetRect.top - overlayRect.height - (SPACING * 2)
+        arrowPosition.top = Math.ceil(overlayRect.height) - ARROW_SPACING - (ARROW_BORDER_SPACING /2)
+      } else {
+        overlayRect.top = targetRect.top + targetRect.height + (SPACING * 2)
+        arrowPosition.top = -ARROW_SPACING
+      }
+
+      if (overlayRect.left < 0) {
+        // if over left edge of screen shift right to 8px of edge
+        // or inline with target (if target is less than 8px from edge)
+        overlayRect.left = Math.min(SPACING, targetRect.left)
+      } else if (overlayDistanceFromRightEdge > documentWidth) {
+        // If over right edge of screen shift left to 8px of edge or inline with target
+        const overRightAmount = overlayDistanceFromRightEdge - documentWidth
+        const targetDistanceFromRight = documentWidth - (targetRect.left + targetRect.width)
+
+        overlayRect.left = overlayRect.left - overRightAmount - Math.min(SPACING, targetDistanceFromRight)
+      }
+
+      arrowPosition.left = targetRect.left - overlayRect.left + targetCenterX - ARROW_SPACING - (ARROW_BORDER_SPACING / 2) // eslint-disable-line max-len
+    } else {
+      const targetCenterY = targetRect.height / 2
+      const overlayCenterY = overlayRect.height / 2
+
+      overlayRect.top = targetRect.top + targetCenterY - overlayCenterY
+      arrowPosition.top = overlayCenterY - ARROW_SPACING - (ARROW_BORDER_SPACING / 2)
+
+      if (placement === 'right') {
+        overlayRect.left = targetRect.left + targetRect.width + (SPACING * 2)
+        arrowPosition.left = -ARROW_SPACING
+      } else {
+        overlayRect.left = targetRect.left - overlayRect.width - (SPACING * 2)
+        arrowPosition.left = overlayRect.width - ARROW_SPACING - (ARROW_BORDER_SPACING / 2)
+      }
+    }
+
+    this.setState({overlayRect: overlayRect, arrowPosition: arrowPosition})
+  }
+
+  render() {
+    const { children, placement } = this.props
+    const { overlayRect } = this.state
+    let computedStyles = styles.root
+
+    if (overlayRect.top) {
+      computedStyles = {
+        ...computedStyles,
+        top: overlayRect.top,
+        left: overlayRect.left
+      }
+    }
+
+    let child = React.Children.only(children)
+
+    child = cloneElement(
+      child,
+      {
+        arrowPosition: this.state.arrowPosition,
+        placement: placement
+      }
+    )
+
+    return (
+      <div style={computedStyles}>
+        { child }
+      </div>
+    )
+
+  }
+}
+
+export default TooltipPosition
```

Comment:
This component is great because it abstracts out the complexity of positioning into a single space. I'm struggling to think of ways that make this less complex without just parting out the `calculatePosition` method into sub-methods. Maybe one method for top/bottom positioning and another for left/right would be good.

Reply:
Ya was thinking about how to make it less complex as well but a lot of variables are shared. When i work on v2 (paragraph tooltips) there will be added logic for left/right positioning for top. Can refactor then if we can wait.